File: app/src/main/java/me/devsaki/hentoid/viewholders/TextItem.java
package me.devsaki.hentoid.viewholders;

import static androidx.core.view.ViewCompat.requireViewById;

import android.graphics.Typeface;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

import com.annimon.stream.function.Consumer;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.drag.IExtendedDraggable;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter.listeners.TouchEventHook;
import com.mikepenz.fastadapter.ui.utils.FastAdapterUIUtils;
import com.mikepenz.fastadapter.utils.DragDropUtil;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.util.StringHelper;
import me.devsaki.hentoid.util.ThemeHelper;

public class TextItem<T> extends AbstractItem<TextItem.TextViewHolder<T>> implements IExtendedDraggable {

    private final String text;
    private final T tag;
    private final boolean centered;
    private final boolean draggable;
    private final boolean reformatCase;
    private final boolean isHighlighted;
    private final ItemTouchHelper touchHelper;


    public TextItem(String text, T tag, boolean centered) {
        this.text = text;
        this.tag = tag;
        this.centered = centered;
        this.draggable = false;
        this.touchHelper = null;
        this.reformatCase = true;
        this.isHighlighted = false;
    }

    public TextItem(String text, T tag, boolean draggable, boolean reformatCase, boolean isHighlighted, ItemTouchHelper touchHelper) {
        this.text = text;
        this.tag = tag;
        this.centered = false;
        this.draggable = draggable;
        this.touchHelper = touchHelper;
        this.reformatCase = reformatCase;
        this.isHighlighted = isHighlighted;
    }

    @Nullable
    @Override
    public T getTag() {
        return tag;
    }

    public String getText() {
        return text;
    }

    private String getDisplayText() {
        return reformatCase ? StringHelper.capitalizeString(text) : text;
    }

    @Override
    public boolean isDraggable() {
        return draggable;
    }

    @Nullable
    @Override
    public ItemTouchHelper getTouchHelper() {
        return touchHelper;
    }

    @NotNull
    @Override
    public TextItem.TextViewHolder<T> getViewHolder(@NotNull View view) {
        return new TextViewHolder<>(view);
    }

    @Nullable
    @Override
    public View getDragView(@NotNull RecyclerView.ViewHolder viewHolder) {
        if (viewHolder instanceof TextViewHolder)
            return ((TextViewHolder<?>) viewHolder).dragHandle;
        else return null;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.item_text;
    }

    @Override
    public int getType() {
        return R.id.text;
    }

    static class TextViewHolder<T> extends FastAdapter.ViewHolder<TextItem<T>> implements IDraggableViewHolder {

        private final View rootView;
        private final TextView title;
        private final ImageView checkedIndicator;
        private final ImageView dragHandle;

        TextViewHolder(View view) {
            super(view);
            rootView = view;
            checkedIndicator = view.findViewById(R.id.checked_indicator);
            dragHandle = view.findViewById(R.id.item_handle);
            title = requireViewById(view, R.id.item_txt);
            int color = ThemeHelper.getColor(view.getContext(), R.color.secondary_light);
            view.setBackground(FastAdapterUIUtils.getSelectablePressedBackground(view.getContext(), FastAdapterUIUtils.adjustAlpha(color, 100), 50, true));
        }


        @Override
        public void bindView(@NotNull TextItem<T> item, @NotNull List<?> list) {
            if (item.draggable) {
                dragHandle.setVisibility(View.VISIBLE);
                DragDropUtil.bindDragHandle(this, item);
            }

            if (item.isSelected()) checkedIndicator.setVisibility(View.VISIBLE);
            else checkedIndicator.setVisibility(View.GONE);

            title.setText(item.getDisplayText());
            if (item.centered) title.setGravity(Gravity.CENTER);

            if (item.isHighlighted) title.setTypeface(null, Typeface.BOLD);
            else title.setTypeface(null, Typeface.NORMAL);
        }

        @Override
        public void unbindView(@NotNull TextItem<T> item) {
            // No specific behaviour to implement
        }

        @Override
        public void onDragged() {
            rootView.setBackgroundColor(ThemeHelper.getColor(rootView.getContext(), R.color.white_opacity_25));
        }

        @Override
        public void onDropped() {
            rootView.setBackgroundColor(ThemeHelper.getColor(rootView.getContext(), R.color.transparent));
        }
    }

    public static class DragHandlerTouchEvent<T> extends TouchEventHook<TextItem<T>> {

        private final Consumer<Integer> action;

        public DragHandlerTouchEvent(@NonNull Consumer<Integer> action) {
            this.action = action;
        }

        @Nullable
        @Override
        public View onBind(@NonNull RecyclerView.ViewHolder viewHolder) {
            return (viewHolder instanceof TextViewHolder<?>) ? ((TextViewHolder<?>) viewHolder).dragHandle : null;
        }

        @Override
        public boolean onTouch(@NonNull View view, @NonNull MotionEvent motionEvent, int position, @NonNull FastAdapter<TextItem<T>> fastAdapter, @NonNull TextItem<T> item) {
            if (motionEvent.getAction() == MotionEvent.ACTION_DOWN) {
                action.accept(position);
                return true;
            }
            return false;
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/viewholders/TextItem.kt
package me.devsaki.hentoid.viewholders

import android.graphics.Typeface
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.drag.IExtendedDraggable
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter.listeners.TouchEventHook
import com.mikepenz.fastadapter.ui.utils.FastAdapterUIUtils
import com.mikepenz.fastadapter.ui.utils.FastAdapterUIUtils.adjustAlpha
import com.mikepenz.fastadapter.ui.utils.FastAdapterUIUtils.getSelectablePressedBackground
import com.mikepenz.fastadapter.utils.DragDropUtil.bindDragHandle
import me.devsaki.hentoid.R
import me.devsaki.hentoid.core.Consumer
import me.devsaki.hentoid.core.requireById
import me.devsaki.hentoid.util.StringHelper
import me.devsaki.hentoid.util.ThemeHelper

class TextItem<T> : AbstractItem<TextItem.ViewHolder<T>>,
    IExtendedDraggable<TextItem.ViewHolder<T>> {
    val text: String
    private val mTag: T?

    private var centered = false
    private var draggable = false
    private var reformatCase = false
    private var highlighted = false
    var isSimple = false

    override var touchHelper: ItemTouchHelper? = null


    constructor(text: String, tag: T, centered: Boolean) : super() {
        this.text = text
        this.mTag = tag
        this.centered = centered
        draggable = false
        touchHelper = null
        reformatCase = true
        highlighted = false
        isSelectable = false
    }

    constructor(text: String, tag: T, reformatCase: Boolean, isSelected: Boolean) : super() {
        this.text = text
        this.mTag = tag
        centered = false
        draggable = false
        touchHelper = null
        this.reformatCase = reformatCase
        highlighted = false
        this.isSelected = isSelected
        isSelectable = true
    }

    constructor(
        text: String,
        tag: T,
        draggable: Boolean,
        reformatCase: Boolean,
        isHighlighted: Boolean,
        centered: Boolean,
        touchHelper: ItemTouchHelper? = null,
        selectable : Boolean = true
    ) : super() {
        this.text = text
        this.mTag = tag
        this.centered = centered
        this.draggable = draggable
        this.touchHelper = touchHelper
        this.reformatCase = reformatCase
        this.highlighted = isHighlighted
        isSelectable = selectable
    }

    fun getObject(): T? {
        return mTag
    }

    private fun getDisplayText(): String {
        return if (reformatCase) StringHelper.capitalizeString(text) else text
    }

    override val isDraggable: Boolean
        get() = draggable

    val isHighlighted: Boolean
        get() = highlighted

    override fun getViewHolder(v: View): ViewHolder<T> {
        return ViewHolder(v)
    }

    override fun getDragView(viewHolder: ViewHolder<T>): View? {
        return viewHolder.dragHandle
    }

    override val layoutRes: Int
        get() = if (isSimple) R.layout.item_text_simple else R.layout.item_text

    override val type: Int
        get() = R.id.text

    class ViewHolder<T> internal constructor(private val root: View) :
        FastAdapter.ViewHolder<TextItem<T>>(root), IDraggableViewHolder {
        private val title: TextView = root.requireById(R.id.item_txt)
        private val checkedIndicator: ImageView? = root.findViewById(R.id.checked_indicator)
        val dragHandle: ImageView? = root.findViewById(R.id.item_handle)

        init {
            val color = ThemeHelper.getColor(root.context, R.color.secondary_light)
            root.background =
                getSelectablePressedBackground(root.context, adjustAlpha(color, 100), 50, true)
        }

        override fun bindView(item: TextItem<T>, payloads: List<Any>) {
            if (item.draggable) {
                dragHandle?.visibility = View.VISIBLE
                @Suppress("UNCHECKED_CAST")
                bindDragHandle(this, item as IExtendedDraggable<RecyclerView.ViewHolder>)
            }

            checkedIndicator?.apply {
                visibility = if (item.isSelected) View.VISIBLE
                else if (item.isSelectable) View.INVISIBLE
                else View.GONE
            }

            title.text = item.getDisplayText()
            if (item.centered) title.gravity = Gravity.CENTER
            if (item.isHighlighted) title.setTypeface(null, Typeface.BOLD)
            else title.setTypeface(null, Typeface.NORMAL)
        }

        override fun unbindView(item: TextItem<T>) {
            // No specific behaviour to implement
        }

        override fun onDragged() {
            root.setBackgroundColor(
                ThemeHelper.getColor(
                    root.context,
                    R.color.white_opacity_25
                )
            )
        }

        override fun onDropped() {
            root.setBackgroundColor(ThemeHelper.getColor(root.context, R.color.transparent))
        }
    }

    class DragHandlerTouchEvent<T>(private val action: Consumer<Int>) :
        TouchEventHook<TextItem<T>>() {
        override fun onBind(viewHolder: RecyclerView.ViewHolder): View? {
            return if (viewHolder is ViewHolder<*>) viewHolder.dragHandle else null
        }

        override fun onTouch(
            v: View,
            event: MotionEvent,
            position: Int,
            fastAdapter: FastAdapter<TextItem<T>>,
            item: TextItem<T>
        ): Boolean {
            if (event.action == MotionEvent.ACTION_DOWN) {
                action.invoke(position)
                return true
            }
            return false
        }
    }
}